xds/internal/balancer/outlierdetection/balancer_test.go

```diff
@@ -287,3 +305,2061 @@ func (errParseConfigBuilder) Name() string {
+
+// TestPicker tests the Picker updates sent upward to grpc from the Outlier
+// Detection Balancer. Two things can trigger a picker update, an
+// UpdateClientConnState call (can flip the no-op config bit that affects
+// Picker) and an UpdateState call (determines the connectivity state sent
+// upward).
+func (s) TestPicker(t *testing.T) {
+	internal.RegisterOutlierDetectionBalancerForTesting()
+	od, tcc := setup(t)
+	defer func() {
+		od.Close()
+		internal.UnregisterOutlierDetectionBalancerForTesting()
+	}()
+
+	od.UpdateClientConnState(balancer.ClientConnState{
+		ResolverState: resolver.State{
+			Addresses: []resolver.Address{
+				{
+					Addr: "address1",
+				},
+			},
+		},
+		BalancerConfig: &LBConfig{ // TODO: S/ variable
+			Interval:           10 * time.Second,
+			BaseEjectionTime:   30 * time.Second,
+			MaxEjectionTime:    300 * time.Second,
+			MaxEjectionPercent: 10,
+			SuccessRateEjection: &SuccessRateEjection{
+				StdevFactor:           1900,
+				EnforcementPercentage: 100,
+				MinimumHosts:          5,
+				RequestVolume:         100,
+			},
+			FailurePercentageEjection: &FailurePercentageEjection{
+				Threshold:             85,
+				EnforcementPercentage: 5,
+				MinimumHosts:          5,
+				RequestVolume:         50,
+			},
+			ChildPolicy: &internalserviceconfig.BalancerConfig{
+				Name:   tcibname,
+				Config: testClusterImplBalancerConfig{},
+			},
+		},
+	})
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	child := od.child.(*testClusterImplBalancer)
+	if err := child.waitForClientConnUpdate(ctx, balancer.ClientConnState{
+		ResolverState: resolver.State{
+			Addresses: []resolver.Address{
+				{
+					Addr: "address1",
+				},
+			},
+		},
+		BalancerConfig: testClusterImplBalancerConfig{},
+	}); err != nil {
+		t.Fatalf("Error waiting for Client Conn update: %v", err)
+	}
+
+	scw, err := od.NewSubConn([]resolver.Address{
+		{
+			Addr: "address1",
+		},
+	}, balancer.NewSubConnOptions{})
+
+	if err != nil {
+		t.Fatalf("error in od.NewSubConn call: %v", err)
+	}
+
+	od.UpdateState(balancer.State{
+		ConnectivityState: connectivity.Ready,
+		Picker: &testutils.TestConstPicker{
+			SC: scw,
+		},
+	})
+
+	// Should forward the connectivity State to Client Conn.
+	select {
+	case <-ctx.Done():
+		t.Fatalf("timeout while waiting for a UpdateState call on the ClientConn")
+	case state := <-tcc.NewStateCh:
+		if state != connectivity.Ready {
+			t.Fatalf("ClientConn received connectivity state %v, want %v", state, connectivity.Ready)
+		}
+	}
+
+	select {
+	case <-ctx.Done():
+		t.Fatalf("timeout while waiting for a UpdateState call on the ClientConn")
+	case picker := <-tcc.NewPickerCh:
+		pi, err := picker.Pick(balancer.PickInfo{})
+		if err != nil {
+			t.Fatalf("Picker.Pick should not have errored")
+		}
+		pi.Done(balancer.DoneInfo{})
+		pi.Done(balancer.DoneInfo{Err: errors.New("some error")})
+		od.mu.Lock()
```

Comment:
I'm generally in favor of testing functionality over testing state. And I would advice that most of these tests be rewritten so that they do not rely on testing any internal state. My experience is that tests written that way are much more maintainable, and offer much better return on investment over time.

Grabbing the lock from the test is definitely not something that I would be comfortable approving.